src/Spago/Watch.hs

```diff
@@ -76,29 +78,37 @@ fileWatchConf watchConfig shouldClear inner = withManagerConf watchConfig $ \man
 
     let onChange event = do
           timeNow <- liftIO getCurrentTime
-
-          rebuilding <- liftIO $ atomically $ do
-            globs                <- readTVar allGlobs
-            (lastTime, lastPath) <- readTVar lastEvent
-
-            let matches glob = Glob.match glob $ Watch.eventPath event
-            -- We should rebuild if at least one of the globs matches the path,
-            -- and the last event either has different path, or has happened
-            -- more than `debounceTime` seconds ago.
-            let shouldRebuild =
-                   any matches globs
-                 && ( lastPath /= Watch.eventPath event
-                   || diffUTCTime timeNow lastTime > debounceTime
-                    )
-
-            when shouldRebuild $ do
-              writeTVar dirtyVar True
-              writeTVar lastEvent (timeNow, Watch.eventPath event)
-
-            pure shouldRebuild
-
-          when rebuilding $ do
-            redisplay $ Just $ "File changed, triggered a build: " <> displayShow (Watch.eventPath event)
+          let eventPath = Watch.eventPath event
+              isPathIgnored =
+                case allowIgnored of
+                  NoAllowIgnored -> isIgnored
+                  DoAllowIgnored -> const (pure False)
+
+          pathIgnored <- isPathIgnored $ Text.pack eventPath
+
+          unless pathIgnored $ do
+            rebuilding <- liftIO $ atomically $ do
+              globs                <- readTVar allGlobs
+              (lastTime, lastPath) <- readTVar lastEvent
+
+              let matches glob = Glob.match glob eventPath
+              -- We should rebuild if at least one of the globs matches the path,
+              -- and the last event either has different path, or has happened
+              -- more than `debounceTime` seconds ago.
+              let shouldRebuild =
```

Comment:
Can we add the `isPathIgnored` check to the `shouldRebuild` instead of wrapping the whole block with it? By doing so we'd keep all the logic about "should we rebuild given a path" in a single place :slightly_smiling_face:

Reply:
That was my initial plan. However, `shouldRebuild` is in the `STM` block. I don't know much about `stm` myself, however AFAIK there is no safe way to execute `IO` actions within `STM`. If there is any other way than the current implementation, I would like to know about it!

Reply:
Oh great point. @klntsky any idea on this? :thinking: